File: pset6/sentimental-credit/credit.py
from cs50 import get_int
from csv import DictReader
import re


# Main function asking for cc number
def main():
    # Get cc number input and turn it into a str
    cc = str(get_int("Number: "))

    # While the string is empty, continue asking for cc number
    while not cc:
        cc = str(get_int("Number: "))

    if luhn(cc):
        print(valid(cc))

    else:
        print("INVALID")


# Function to run Luhn's algorithm
def luhn(string):
    # Set variables needed for the algorithm
    penultimate = len(string) - 2
    sum = 0

    # Multiply every other digit by 2 then add each digit,
    # starting from second last digit
    for i in range(penultimate, -1, -2):
        temp = str(int(string[i]) * 2)
        if len(temp) > 1:
            sum += int(temp[0]) + int(temp[1])
        else:
            sum += int(temp)
        # print(f"multiply: {sum}")

    # Add the sum to the sum of the other digits
    for j in range(penultimate + 1, -1, -2):
        sum += int(string[j])
    #     print(f"add: {sum}")

    # print(sum)

    if sum % 10 == 0:
        return True
    else:
        return False

# # Function to check if number belongs to a valid cc company
# def valid(string):
#     corps = []

#     # Create a dict using the csv file
#     with open('cc.csv', 'r') as file:
#         reader = DictReader(file)

#         for row in reader:
#             corps.append(row)

#     # Cycle through each code seeing if it matches input
#     for corp in corps:
#         pattern = re.compile(corp["code"])
#         result = re.match(pattern, string)

#         # If a match is found, print the name of cc company
#         if result:
#             return corp["cc"].upper()
#             # print(f"{company}")
#             # break
#     else:
#         # If no match was found, print invalid
#         return "INVALID"
#         # print("INVALID")


# Function to check if number belongs to a valid cc company
def valid(string):
    # Create a dict of all the corps and their codes
    corps = {"34": "amex", "37": "amex", "51": "mastercard",
             "52": "mastercard", "53": "mastercard", "54": "mastercard",
             "55": "mastercard", "4": "visa"}

    # Cycle through each code seeing if it matches input
    for code, corp in corps.items():
        pattern = re.compile(code)
        result = re.match(pattern, string)

        # If a match is found, print the name of cc company
        if result:
            if corp == "amex" and len(string) == 15:
                return corp.upper()
            if corp == "mastercard" and len(string) == 16:
                return corp.upper()
            if corp == "visa" and len(string) == 13 or len(string) == 16:
                return corp.upper()
    else:
        # If no matches, return invalid
        return "INVALID"


main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// Program to validate credit card numbers

// Functions that will validate if credit card is valid for relevant company
bool amex(long cc);
bool mastercard(long cc);
bool visa(long cc);

int main(void)
{
    long cc = get_long("What's your credit card number? ");

    // If functions are true, print out type of card. Otherwise, print invalid
    if (amex(cc))
    {
        printf("AMEX\n");
    }
    else if (mastercard(cc))
    {
        printf("MASTERCARD\n");
    }
    else if (visa(cc))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Function validate amex
bool amex(long cc)
{
    int digits = 0;
    long sum = 0;
    long first_two_digits = 0;
    long i = 0;
    long j = 0;

    // Multiply every other digit by 2, starting with second-to-last digit
    for (i = 100; i <= cc * 10; i = i * 100)
    {
        if (((cc % i) / (i / 10)) < 5)
        {
            sum = sum + (((cc % i) / (i / 10)) * 2);
        }
        else
        {
            int test = (cc % i) / (i / 10);
            sum = sum + (((cc % i) / (i / 10) * 2) / 10) + (((cc % i) / (i / 10) * 2) % 10);
        }
        digits++;
    }

    // Add the sum of every other digit starting with the last
    for (j = 10; j <= cc * 10; j = j * 100)
    {
        sum = sum + ((cc % j) / (j / 10));
        digits++;
    }

    // Find if modulo 10 of sum of all digits is zero
    if (sum % 10 != 0)
    {
        return false;
    }

    // Find and validate the first two digits
    if (digits > 1)
    {
        if (i > j)
        {
            first_two_digits = cc / (i / 10000);
        }
        else
        {
            first_two_digits = cc / (j / 10000);
        }
    }

    if (first_two_digits != 34 && first_two_digits != 37)
    {
        return false;
    }

    // Validate number of digits
    if (digits != 15)
    {
        return false;
    }

    return true;
}

// Function to validate mastercards
bool mastercard(long cc)
{
    int digits = 0;
    long sum = 0;
    long first_two_digits = 0;
    long i = 0;
    long j = 0;

    // Multiply every other digit by 2, starting with second-to-last digit
    for (i = 100; i <= cc * 10; i = i * 100)
    {
        if (((cc % i) / (i / 10)) < 5)
        {
            sum = sum + (((cc % i) / (i / 10)) * 2);
        }
        else
        {
            int test = (cc % i) / (i / 10);
            sum = sum + (((cc % i) / (i / 10) * 2) / 10) + (((cc % i) / (i / 10) * 2) % 10);
        }
        digits++;
    }

    // Add the sum of every other digit starting with the last
    for (j = 10; j <= cc * 10; j = j * 100)
    {
        sum = sum + ((cc % j) / (j / 10));
        digits++;
    }

    // Find if modulo 10 of sum of all digits is zero
    if (sum % 10 != 0)
    {
        return false;
    }

    // Find and validate the first two digits
    if (digits > 1)
    {
        if (i > j)
        {
            first_two_digits = cc / (i / 10000);
        }
        else
        {
            first_two_digits = cc / (j / 10000);
        }
    }

    if (first_two_digits > 55 || first_two_digits < 51)
    {
        return false;
    }

    // Validate number of digits
    if (digits != 16)
    {
        return false;
    }

    return true;
}

// Function to validate visa cards
bool visa(long cc)
{
    int digits = 0;
    long sum = 0;
    long first_digit = 0;
    long i = 0;
    long j = 0;

    // Multiply every other digit by 2, starting with second-to-last digit
    for (i = 100; i <= cc * 10; i = i * 100)
    {
        if (((cc % i) / (i / 10)) < 5)
        {
            sum = sum + (((cc % i) / (i / 10)) * 2);
        }
        else
        {
            int test = (cc % i) / (i / 10);
            sum = sum + (((cc % i) / (i / 10) * 2) / 10) + (((cc % i) / (i / 10) * 2) % 10);
        }
        digits++;
    }

    // Add the sum of every other digit starting with the last
    for (j = 10; j <= cc * 10; j = j * 100)
    {
        sum = sum + ((cc % j) / (j / 10));
        digits++;
    }

    // Find if modulo 10 of sum of all digits is zero
    if (sum % 10 != 0)
    {
        return false;
    }

    // Find and validate the first digit
    if (digits > 1)
    {
        if (i > j)
        {
            first_digit = cc / (i / 1000);
        }
        else
        {
            first_digit = cc / (j / 1000);
        }
    }
    if (first_digit != 4)
    {
        return false;
    }

    // Validate number of digits
    if (digits != 13 && digits != 16)
    {
        return false;
    }

    return true;
}